File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math


# 프로그래머스 92341 주차 요금 계산
def solution(fees, records):
    answer = []
    r_time = {}
    
    # 차량별 누적 주차 시간 계산
    for r in records:
        _r = r.split()
        _r[0] = int(_r[0].split(':')[0]) * 60 + int(_r[0].split(':')[1])

        # 여기서 IN 의 경우 (-), OUT 의 경우 (+) 를 해준다.
        # OUT - IN = 주차한 시간
        if not r_time.get(_r[1]):
            if _r[2] == 'IN':
                r_time[_r[1]] = -_r[0]
            else:
                r_time[_r[1]] = _r[0]
        else:
            if _r[2] == 'IN':
                r_time[_r[1]] -= _r[0]
            else:
                r_time[_r[1]] += _r[0]

    # 누적 주차 시간을 이용하여 차량별 요금 계산
    for k, v in sorted(r_time.items(), key=lambda x: x[0]):
        # value(누적 주차 시간) 가 0 이하라면 23:59 까지 주차한 것이므로 그만큼 더해준다.
        if v <= 0:
            v += 1439

        if v > fees[0]:
            answer.append(fees[1] + math.ceil((v - fees[0]) / fees[2]) * fees[3])
        else:
            answer.append(fees[1])

    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;
import java.util.stream.*;

class TimeConverter {
    
    TimeConverter() {}
    
    int convert(String time) {
        String[] splited = time.split(":");
        return Integer.parseInt(splited[0]) * 60 + Integer.parseInt(splited[1]);
    }
}

class Parking {
    private int in, out;
    
    public Parking(int in) {
        this.in = in;
    }
    
    public void setOut(int out) {
        this.out = out;
    }
    
    public int getIn() {
        return this.in;
    }
    
    public int getOut() {
        return this.out;
    }
    
    public int getParkingTime() {
        int tmp = this.out;
        if (tmp == 0) tmp = 1439;
        return tmp - this.in;
    }
}

class Solution {
    final String IN = "IN", OUT = "OUT";
    
    public int[] solution(int[] fees, String[] records) {
        Map<String, List<Parking>> parkingMap = new HashMap<>();
        TimeConverter tc = new TimeConverter();
        
        for (String record : records) {
            String[] info = record.split(" ");
            if (info[2].equals(IN)) {
                parkingMap.putIfAbsent(info[1], new ArrayList<>());
                parkingMap.get(info[1]).add(new Parking(tc.convert(info[0])));
            } else {
                int size = parkingMap.get(info[1]).size();
                parkingMap.get(info[1]).get(size - 1).setOut(tc.convert(info[0]));
            }
        }
        
        List<String> carIds = new ArrayList<>(parkingMap.keySet());
        carIds.sort(Comparator.naturalOrder());
        
        List<Integer> answer = new ArrayList<>();
        for (String carId : carIds) {
            List<Parking> logs = parkingMap.get(carId);
            
            int accumulated = 0;
            for (Parking log : logs) accumulated += log.getParkingTime();
            
            if (accumulated <= fees[0]) answer.add(fees[1]);
            else answer.add(fees[1] + (int) Math.ceil((accumulated - fees[0]) * 1.0 / fees[2]) * fees[3]);
        }
        
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}